src/runtime_src/core/pcie/windows/device_windows.cpp

```diff
@@ -835,6 +835,83 @@ struct rom
   }
 };
 
+struct data_retention
+{
+  using result_type = uint32_t;
+  using value_type = uint32_t;
+  
+  static result_type
+  user_get(const xrt_core::device* device)
+  {
+	  return 0;
+  }
+
+  static result_type
+  mgmt_get(const xrt_core::device* device)
+  {
+    auto init_ret = [](const xrt_core::device* dev) {
+    uint32_t ret;
+    mgmtpf::get_data_retention(dev->get_mgmt_handle(), &ret);
+    return ret;
+    };
+
+    static std::map<const xrt_core::device*, uint32_t> info_map;
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> lk(mutex);
+    auto it = info_map.find(device);
+    if (it == info_map.end()) {
+      auto ret = info_map.emplace(device,init_ret(device));
+      it = ret.first;
+    }
+
+    return (*it).second;
+  }
+
+  static void
+  user_put(const xrt_core::device* device, value_type)
+  {
+    throw std::runtime_error("device data retention query is not implemented on user windows");
+  }
+  static void
+    mgmt_put(const xrt_core::device* device, value_type val)
+  {
+    mgmtpf::set_data_retention(device->get_mgmt_handle(), val);
```

Comment:
Is this thread safe, or does it not matter, e.g. will there be one thread only?

Reply:
You are right, this isn't thread-safe. I'll add code to make it so. I'm not sure if it matters or not, but it doesn't hurt to take precautions :)